### PR TITLE
RFR: Added Badge Endpoint

### LINKED ISSUE #187: BE: Badges Endpoint
**Is your feature request related to a problem? Please describe.**
Right now frontend needs a way to get badges in the form of 
```
{
  "badgesEarned": [],
  "badgesInProgress": [
  {
    "type": 'perfect scores', (or 'completed quizzes' or 'completed books')
    "currentCount": 3,
    "targetCount": 5 (only reflects the upcoming target count for next badge)
  }
] (if none of the badges are in progress, then this array is empty)
}
```

### PR DESCRIPTION
**Pull request by Davidw1339**
## Status: 
:rocket: Ready

## Description
Added in the badges endpoint. Currently a badge is a static model in memory (not a SQL table). All the data is 'technically' hardcoded such as the text/descriptions/graphics. 

Calling /badges will now return data in the form similar to 
```
{
   badgesEarned: [],
   badgesInProgress: [
      {
         type: 'perfect scores', (or 'completed quizzes' or 'completed books')
         currentCount: 3,
         targetCount: 5 (only reflects the upcoming target count for next badge)
      }
   ] (if none of the badges are in progress, then this array is empty)
}

New badges are returned as data from the /quiz_result POST endpoint
```

Related issues: fixes #187 

## Screenshots
**Note** Not all data retrieved from badges endpoint is shown
image

### CODE REVIEW

File: backend/api/models/Badge.py [hunk 1]
```diff
@@ -1,22 +1,21 @@
 import pdb
```
Comment (micleo2) [inline backend/api/models/Badge.py]: nitpick: delete?

File: backend/api/models/Badge.py [hunk 2]
```diff
@@ -25,63 +24,200 @@ def serialize_to_json(self):
...
+                    earned_badges.append(badge)
+        return earned_badges
+
+    def get_progress(self, user, quiz_result):
+        return 0
+
+
+def num_quiz_attempts(user):
+    quiz_set = set()
+    for quiz_result in user.attempted_quizzes:
```
Comment (micleo2) [inline backend/api/models/Badge.py]: why can't you return len(user.attempted_questions)? Will there be duplicate items in attempted quizzes?

File: backend/api/views/user.py [hunk 3]
```diff
@@ -107,3 +107,21 @@ def user_info():
...
+
+# returns all in progress badges and already earned badges
+@user.route("/badges", methods=["GET"])
+@authenticated_route
+def badges(user_id):
+    user = User.query.get(user_id)
+    if user is None:
+        return create_response(
+            message="User not found", status=401, data={"status": "fail"}
+        )
```
Comment (micleo2) [inline backend/api/views/user.py]: :lock: good work

File: backend/api/models/Badge.py [hunk 4]
```diff
@@ -25,63 +24,200 @@ def serialize_to_json(self):
...
+        )
+    for badge in new_badges:
+        user.badges.append(badge.id)
+
+    return [badge.serialize_to_json() for badge in new_badges]
 
-    return new_badges
+
+def get_user_badges(user):
+    return [all_badge_map[badge_id] for badge_id in user.badges]
```
Comment (micleo2) [inline backend/api/models/Badge.py]: succicnt list comphrension :heavy_check_mark:

File: backend/api/models/Badge.py [hunk 5]
```diff
@@ -25,63 +24,200 @@ def serialize_to_json(self):
...
+        if next_badge_progress_or_none is not None:
+            next_badge_progress, next_badge = next_badge_progress_or_none
+            category_progress = {
+                "type": category.name,
+                "currentCount": next_badge_progress,
+                "targetCount": next_badge.quantity,
+            }
+            category_progress.update(next_badge.serialize_to_json())
+            progress.append(category_progress)
+    return progress
```
Comment (micleo2) [inline backend/api/models/Badge.py]: unexpected change :lock:

File: backend/api/views/user.py (hunk 3 above)
Comment (Davidw1339) [reply backend/api/views/user.py]: thank u thank u

File: backend/api/models/Badge.py (hunk 5 above)
Comment (Davidw1339) [reply backend/api/models/Badge.py]: unexpected comment :scroll:

**Pull request merged**